File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/shadcn-input-group-no-raw-controls.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findJsxAttribute } from "../../utils/find-jsx-attribute.js";
import { flattenJsxName } from "../../utils/flatten-jsx-name.js";
import { getJsxPropStringValue } from "../../utils/get-jsx-prop-string-value.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { resolveShadcnUiComponentName } from "../../utils/resolve-shadcn-ui-component-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";

const INPUT_GROUP_MODULE_PATTERN =
  /(?:^|\/)ui\/(?:.*\/)?input-group$|^\.\.?\/(?:.*\/)?input-group$/;
const CANONICAL_INPUT_GROUP_PARTS: ReadonlySet<string> = new Set([
  "InputGroupAddon",
  "InputGroupInput",
  "InputGroupTextarea",
]);

interface RawControlContract {
  readonly uiComponent: string;
  readonly uiModuleSourcePattern: RegExp;
  readonly nativeTag: string;
  readonly replacement: string;
}

// InputGroup owns the group's border, focus ring, and error ring through its
// `data-slot="input-group-control"` child selectors. A raw control dropped
// directly inside keeps its own border and focus ring, so the group renders
// a control-inside-a-control double frame and the group ring never activates.
const RAW_CONTROL_CONTRACTS: ReadonlyArray<RawControlContract> = [
  {
    uiComponent: "Input",
    uiModuleSourcePattern: /(?:^|\/)ui\/(?:.*\/)?input$|^\.\.?\/(?:.*\/)?input$/,
    nativeTag: "input",
    replacement: "InputGroupInput",
  },
  {
    uiComponent: "Textarea",
    uiModuleSourcePattern: /(?:^|\/)ui\/(?:.*\/)?textarea$|^\.\.?\/(?:.*\/)?textarea$/,
    nativeTag: "textarea",
    replacement: "InputGroupTextarea",
  },
  {
    uiComponent: "Button",
    uiModuleSourcePattern: /(?:^|\/)ui\/(?:.*\/)?button$|^\.\.?\/(?:.*\/)?button$/,
    nativeTag: "button",
    replacement: "InputGroupButton inside an InputGroupAddon",
  },
];

const collectDirectRenderedElements = (
  children: ReadonlyArray<EsTreeNode>,
  elements: Array<EsTreeNodeOfType<"JSXElement">>,
): void => {
  for (const child of children) {
    if (isNodeOfType(child, "JSXElement")) {
      elements.push(child);
    } else if (isNodeOfType(child, "JSXFragment")) {
      collectDirectRenderedElements(child.children, elements);
    } else if (isNodeOfType(child, "JSXExpressionContainer")) {
      collectDirectRenderedExpressionElements(child.expression, elements);
    }
  }
};

// Follows rendering structure (fragments, conditionals, logical guards,
// arrays) without crossing into another element's children — a control
// nested inside an InputGroupAddon is that part's concern, not the group's.
const collectDirectRenderedExpressionElements = (
  rawExpression: EsTreeNode,
  elements: Array<EsTreeNodeOfType<"JSXElement">>,
): void => {
  const expression = stripParenExpression(rawExpression);
  if (isNodeOfType(expression, "JSXElement")) {
    elements.push(expression);
    return;
  }
  if (isNodeOfType(expression, "JSXFragment")) {
    collectDirectRenderedElements(expression.children, elements);
    return;
  }
  if (isNodeOfType(expression, "ConditionalExpression")) {
    collectDirectRenderedExpressionElements(expression.consequent, elements);
    collectDirectRenderedExpressionElements(expression.alternate, elements);
    return;
  }
  if (isNodeOfType(expression, "LogicalExpression")) {
    collectDirectRenderedExpressionElements(expression.right, elements);
    if (expression.operator !== "&&") {
      collectDirectRenderedExpressionElements(expression.left, elements);
    }
    return;
  }
  if (isNodeOfType(expression, "ArrayExpression")) {
    for (const element of expression.elements) {
      if (element && !isNodeOfType(element, "SpreadElement")) {
        collectDirectRenderedExpressionElements(element, elements);
      }
    }
  }
};

const getRawControlContract = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
  context: RuleContext,
): RawControlContract | null => {
  const flattenedName = flattenJsxName(openingElement.name);
  for (const contract of RAW_CONTROL_CONTRACTS) {
    if (
      flattenedName === contract.nativeTag ||
      resolveShadcnUiComponentName(openingElement.name, contract.uiModuleSourcePattern, context) ===
        contract.uiComponent
    ) {
      return contract;
    }
  }
  return null;
};

const isCanonicalInputGroupPart = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
  context: RuleContext,
): boolean => {
  const componentName = resolveShadcnUiComponentName(
    openingElement.name,
    INPUT_GROUP_MODULE_PATTERN,
    context,
  );
  return componentName !== null && CANONICAL_INPUT_GROUP_PARTS.has(componentName);
};

// A hidden input renders nothing, so it cannot double-frame the group.
const isHiddenInput = (openingElement: EsTreeNodeOfType<"JSXOpeningElement">): boolean => {
  const typeAttribute = findJsxAttribute(openingElement.attributes, "type");
  return typeAttribute !== undefined && getJsxPropStringValue(typeAttribute) === "hidden";
};

export const shadcnInputGroupNoRawControls = defineRule({
  id: "shadcn-input-group-no-raw-controls",
  title: "Raw control directly inside InputGroup",
  severity: "warn",
  category: "Correctness",
  requires: ["shadcn"],
  matchByOccurrence: true,
  recommendation:
    "Compose InputGroup from its own parts (InputGroupInput, InputGroupTextarea, and InputGroupAddon with InputGroupButton) so the group owns one border, focus ring, and error state.",
  create: (context: RuleContext) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      if (
        resolveShadcnUiComponentName(node.name, INPUT_GROUP_MODULE_PATTERN, context) !==
        "InputGroup"
      ) {
        return;
      }
      const groupElement = node.parent;
      if (!groupElement || !isNodeOfType(groupElement, "JSXElement")) return;
      const directChildElements: Array<EsTreeNodeOfType<"JSXElement">> = [];
      collectDirectRenderedElements(groupElement.children, directChildElements);
      if (
        !directChildElements.some((childElement) =>
          isCanonicalInputGroupPart(childElement.openingElement, context),
        )
      ) {
        return;
      }
      for (const childElement of directChildElements) {
        const contract = getRawControlContract(childElement.openingElement, context);
        if (!contract || isHiddenInput(childElement.openingElement)) continue;
        const childName = flattenJsxName(childElement.openingElement.name) ?? contract.nativeTag;
        context.report({
          node: childElement.openingElement,
          message: `This ${childName} sits directly inside InputGroup, so it keeps its own border and focus ring and the group's shared focus and error styling never applies. Use ${contract.replacement} instead.`,
        });
      }
    },
  }),
});

File: native/oxlint/rules/shadcn-input-group-no-raw-controls.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{AstNode, context::LintContext, rule::Rule};

const CANONICAL_INPUT_GROUP_PARTS: [&str; 3] = [
    "InputGroupAddon",
    "InputGroupInput",
    "InputGroupTextarea",
];

#[derive(Debug, Default, Clone)]
pub struct ShadcnInputGroupNoRawControls;

declare_oxc_lint!(
    /// Disallow raw controls directly inside shadcn InputGroup.
    ShadcnInputGroupNoRawControls,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow raw controls directly inside shadcn InputGroup.",
);

impl Rule for ShadcnInputGroupNoRawControls {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if !has_capability_or_unspecified(ctx, "shadcn") {
            return;
        }
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        if resolve_shadcn_component_name(&opening_element.name, "input-group", ctx).as_deref()
            != Some("InputGroup")
        {
            return;
        }
        let parent = ctx.nodes().parent_node(node.id());
        let AstKind::JSXElement(group_element) = parent.kind() else {
            return;
        };
        let mut direct_child_elements = Vec::new();
        visit_static_jsx_children(
            &group_element.children,
            &mut |child_element| {
                direct_child_elements.push(child_element);
                false
            },
            &mut || {},
            &mut || {},
        );
        if !direct_child_elements.iter().any(|child_element| {
            resolve_shadcn_component_name(
                &child_element.opening_element.name,
                "input-group",
                ctx,
            )
            .is_some_and(|component_name| {
                CANONICAL_INPUT_GROUP_PARTS.contains(&component_name.as_str())
            })
        }) {
            return;
        }
        for child_element in direct_child_elements {
            let child_opening_element = &child_element.opening_element;
            let Some((native_tag, replacement)) = raw_control_contract(child_opening_element, ctx)
            else {
                continue;
            };
            if find_jsx_attribute(child_opening_element, "type")
                .and_then(|attribute| get_string_literal_attribute_value(attribute))
                == Some("hidden")
            {
                continue;
            }
            let child_name = child_opening_element
                .name
                .span()
                .source_text(ctx.source_text());
            let child_name = if child_name.is_empty() {
                native_tag
            } else {
                child_name
            };
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "This {child_name} sits directly inside InputGroup, so it keeps its own border and focus ring and the group's shared focus and error styling never applies. Use {replacement} instead."
                ))
                .with_label(child_opening_element.span),
            );
        }
    }
}

fn raw_control_contract<'a>(
    opening_element: &oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &LintContext<'a>,
) -> Option<(&'static str, &'static str)> {
    let element_name = &opening_element.name;
    if matches!(
        element_name,
        oxc_ast::ast::JSXElementName::Identifier(identifier) if identifier.name == "input"
    )
        || resolve_shadcn_component_name(element_name, "input", ctx).as_deref() == Some("Input")
    {
        return Some(("input", "InputGroupInput"));
    }
    if matches!(
        element_name,
        oxc_ast::ast::JSXElementName::Identifier(identifier) if identifier.name == "textarea"
    )
        || resolve_shadcn_component_name(element_name, "textarea", ctx).as_deref()
            == Some("Textarea")
    {
        return Some(("textarea", "InputGroupTextarea"));
    }
    if matches!(
        element_name,
        oxc_ast::ast::JSXElementName::Identifier(identifier) if identifier.name == "button"
    )
        || resolve_shadcn_component_name(element_name, "button", ctx).as_deref() == Some("Button")
    {
        return Some((
            "button",
            "InputGroupButton inside an InputGroupAddon",
        ));
    }
    None
}
